// service-infrastructure/health/ready.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var ReadyService_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReadyService = exports.NOT_READY = void 0;
const common_1 = require("@nestjs/common");
const terminus_1 = require("@nestjs/terminus");
const configuration_service_1 = require("../common/config/configuration.service");
const mculogger_service_1 = require("../common/logs/mculogger.service");
const memory_rss_service_1 = require("./memory/memory.rss.service");
const disk_service_1 = require("./disk/disk.service");
const base_health_service_1 = require("./base.health.service");
const discovery_service_1 = require("./discovery/discovery.service");
const health_def_1 = require("./health.def");
const cpu_service_1 = require("./cpu/cpu.service");
const shutdown_probe_1 = require("./shutdown.probe");
exports.NOT_READY = "false";
let ReadyService = ReadyService_1 = class ReadyService extends base_health_service_1.BaseHealthService {
    constructor(memoryRssService, diskService, discoveryService, cpuService, configService, logger, health, shutdown) {
        super(configService, logger, health);
        this.memoryRssService = memoryRssService;
        this.diskService = diskService;
        this.discoveryService = discoveryService;
        this.cpuService = cpuService;
        this.shutdown = shutdown;
        super.setApplicationName(ReadyService_1.name);
        super.setProbsList([memoryRssService, diskService, discoveryService], health_def_1.ProbType.READY);
    }
    setProbs(healthProbs, override = false) {
        super.setProbsList(healthProbs, health_def_1.ProbType.READY, override);
    }
    async healthCheck(readFromCache = true) {
        if (process.env.READY && process.env.READY == exports.NOT_READY) {
            return this.shutdown.check();
        }
        return await super.healthCheck(readFromCache);
    }
};
ReadyService = ReadyService_1 = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [memory_rss_service_1.MemoryRssService,
        disk_service_1.DiskService,
        discovery_service_1.DiscoveryService,
        cpu_service_1.CpuService,
        configuration_service_1.ConfigurationService,
        mculogger_service_1.MculoggerService,
        terminus_1.HealthCheckService,
        shutdown_probe_1.ShutdownProbe])
], ReadyService);
exports.ReadyService = ReadyService;
