// service-infrastructure/health/memory/memory.rss.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MemoryRssService = void 0;
const common_1 = require("@nestjs/common");
const terminus_1 = require("@nestjs/terminus");
const configuration_service_1 = require("../../common/config/configuration.service");
const health_def_1 = require("../health.def");
let MemoryRssService = class MemoryRssService extends terminus_1.MemoryHealthIndicator {
    constructor(configService) {
        super();
        this.configService = configService;
    }
    async check() {
        return await this.isHealthy('memory_rss');
    }
    async isHealthy(key) {
        let memoryThreshold = (health_def_1.ProbType.READY === this.probType) ? (Number(this.configService.get('health.readyThreshold.maxMemory', 1024))) : (Number(this.configService.get('health.healthyThreshold.maxMemory', 1024)));
        return await super.checkRSS('memory_rss', memoryThreshold * 1024 * 1024);
    }
    setProbType(probType = health_def_1.ProbType.HEALTH) {
        this.probType = probType;
    }
};
MemoryRssService = __decorate([
    (0, common_1.Injectable)({ scope: common_1.Scope.TRANSIENT }),
    __metadata("design:paramtypes", [configuration_service_1.ConfigurationService])
], MemoryRssService);
exports.MemoryRssService = MemoryRssService;
