// service-infrastructure/common/config/configuration/discoveryConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = () => ({
    discovery: {
        type: "dns",
        enable: true,
        services: {
            mediaGw: {
                name: "media-gateway",
                ttl: 2000
            },
            mediaServer: {
                name: "media-server",
            },
            stats: {
                name: "mcu-stats-service",
                ttl: 90000
            },
            recorder: {
                name: "mcu-recorder-controller",
                ttl: 6000
            },
            sipGwNotify: {
                name: "sip-notify-internal",
                ttl: 2000
            },
            sipGw: {
                name: "sip-gateway-internal",
                ttl: 2000
            },
            monitor: {
                name: "mcu-monitor-service",
                ttl: 3000
            },
            etcd: {
                name: "etcd",
            }
        }
    }
});
