// service-infrastructure/dd-metrics/hotshouts.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataDogModule = void 0;
const common_1 = require("@nestjs/common");
const nestjs_hot_shots_1 = require("nestjs-hot-shots");
const metrics_service_1 = require("./metrics.service");
let DataDogModule = class DataDogModule {
};
DataDogModule = __decorate([
    (0, common_1.Module)({
        imports: [
            nestjs_hot_shots_1.HotShotsModule.forRoot({
                host: process.env.DD_AGENT_HOST || process.env.HBIND,
                port: 8125,
                mock: false,
                protocol: "udp",
                errorHandler: err => (error) => console.log("Socket errors caught here: ", error),
                globalTags: { env: process.env.NODE_ENV, service: process.env.SERVICE_NAME || 'service.infrastructure' }
            })
        ],
        providers: [metrics_service_1.MetricsService],
        exports: [metrics_service_1.MetricsService]
    })
], DataDogModule);
exports.DataDogModule = DataDogModule;
