// mcu-discovery/lib/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SERVICE_NAME = "SERVICE_NAME";
exports.SERVICE_PORT = "SERVICE_PORT";
exports.SERVICE_TAGS = "SERVICE_TAGS";
exports.SERVICE_CHECK_DEREGISTER_AFTER = "SERVICE_CHECK_DEREGISTER_AFTER";
exports.SERVICE_CHECK_HTTP = "SERVICE_CHECK_HTTP";
exports.SERVICE_CHECK_TYPE = "SERVICE_CHECK_TYPE";
exports.SERVICE_CHECK_INTERVAL = "SERVICE_CHECK_INTERVAL";
var ServiceType;
(function (ServiceType) {
    ServiceType[ServiceType["sipgw"] = 11] = "sipgw";
    ServiceType[ServiceType["msml"] = 12] = "msml";
    ServiceType[ServiceType["janusgw"] = 13] = "janusgw";
    ServiceType[ServiceType["janus"] = 14] = "janus";
    ServiceType[ServiceType["stats"] = 15] = "stats";
    ServiceType[ServiceType["mediagw"] = 16] = "mediagw";
    ServiceType[ServiceType["media"] = 17] = "media";
    ServiceType[ServiceType["recordermng"] = 18] = "recordermng";
    ServiceType[ServiceType["recorderdrv"] = 19] = "recorderdrv";
    ServiceType[ServiceType["mediagwdev"] = 20] = "mediagwdev";
})(ServiceType = exports.ServiceType || (exports.ServiceType = {}));
//function overrideRegisterService(consul : any){
// consul.prototype.register = function(opts, callback) {
//     if (typeof opts === 'string') {
//       opts = { name: opts };
//     }
//     opts = utils.normalizeKeys(opts);
//     opts = utils.defaults(opts, this.consul._defaults);
//     var req = {
//       name: 'agent.service.register',
//       path: '/agent/service/register',
//       type: 'json',
//       body: {},
//     };
//     if (!opts.name) {
//       return callback(this.consul._err(errors.Validation('name required'), req));
//     }
//     req.body.Name = opts.name;
//     if (opts.id) req.body.ID = opts.id;
//     if (opts.tags) req.body.Tags = opts.tags;
//     if (opts.meta) req.body.Meta =  opts.meta;
//     if (opts.hasOwnProperty('address')) req.body.Address = opts.address;
//     if (opts.hasOwnProperty('port')) req.body.Port = opts.port;
//     try {
//       if (Array.isArray(opts.checks)) {
//         req.body.Checks = opts.checks.map(utils.createServiceCheck);
//       } else if (opts.check) {
//         req.body.Check = utils.createServiceCheck(opts.check);
//       }
//     } catch (err) {
//       return callback(this.consul._err(errors.Validation(err.message), req));
//     }
//     utils.options(req, opts);
//     this.consul._put(req, utils.empty, callback);
//   };
// }
