// nestjs-hot-shots/dist/hot-shots.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var HotShotsModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.HotShotsModule = void 0;
const common_1 = require("@nestjs/common");
const hot_shots_constants_1 = require("./hot-shots.constants");
const hot_shots_1 = require("hot-shots");
let HotShotsModule = HotShotsModule_1 = class HotShotsModule {
    static forRoot(options) {
        const StatsDProvider = {
            provide: hot_shots_1.StatsD,
            useValue: new hot_shots_1.StatsD(options)
        };
        return {
            module: HotShotsModule_1,
            providers: [
                {
                    provide: hot_shots_constants_1.HOT_SHOTS_MODULE_OPTIONS,
                    useValue: options
                },
                StatsDProvider
            ],
            exports: [StatsDProvider]
        };
    }
    static forRootAsync(options) {
        const StatsDFactoryProvider = {
            provide: hot_shots_1.StatsD,
            useFactory: (options) => new hot_shots_1.StatsD(options),
            inject: [hot_shots_constants_1.HOT_SHOTS_MODULE_OPTIONS]
        };
        return {
            module: HotShotsModule_1,
            imports: options.imports,
            providers: this.createAsyncProviders(options).concat(StatsDFactoryProvider),
            exports: [StatsDFactoryProvider]
        };
    }
    static createAsyncProviders(options) {
        if (options.useExisting || options.useFactory) {
            return [this.createAsyncOptionsProvider(options)];
        }
        return [
            this.createAsyncOptionsProvider(options),
            {
                provide: options.useClass,
                useClass: options.useClass
            }
        ];
    }
    static createAsyncOptionsProvider(options) {
        if (options.useFactory) {
            return {
                provide: hot_shots_constants_1.HOT_SHOTS_MODULE_OPTIONS,
                useFactory: (...args) => __awaiter(this, void 0, void 0, function* () { return yield options.useFactory(...args); }),
                inject: options.inject || []
            };
        }
        return {
            provide: hot_shots_constants_1.HOT_SHOTS_MODULE_OPTIONS,
            useFactory: (optionsFactory) => __awaiter(this, void 0, void 0, function* () { return yield optionsFactory.createHotShotsOptions(); }),
            inject: [options.useExisting || options.useClass]
        };
    }
};
HotShotsModule = HotShotsModule_1 = __decorate([
    (0, common_1.Global)(),
    (0, common_1.Module)({})
], HotShotsModule);
exports.HotShotsModule = HotShotsModule;
